lxd/network/network_utils.go

```diff
@@ -639,6 +639,111 @@ func randomSubnetV6() (string, error) {
 	return "", fmt.Errorf("Failed to automatically find an unused IPv6 subnet, manual configuration required")
 }
 
+// noAvailableAddressErr is used by randomAddressInSubnet to indicate that the subnet was exhausted while searching for
+// IP addresses.
+type noAvailableAddressErr struct {
+	error
+}
+
+// Unwrap implements xerrors.Unwrap for noAvailableAddressErr.
+func (e noAvailableAddressErr) Unwrap() error {
+	return e.error
+}
+
+// randomAddressInSubnet finds a random address within a given subnet. If given, the validate function is called for
+// each randomly generated value. Each time validate returns an error, another IP address will be generated until all
+// IPs have been exhausted or the context is cancelled.
+func randomAddressInSubnet(ctx context.Context, ipnet net.IPNet, validate func(net.IP) error) (net.IP, error) {
+	ones, size := ipnet.Mask.Size()
+	subnetExponent := size - ones
+
+	// networkAddress is the first address in the subnet.
+	var networkAddress net.IP
+	if ipnet.IP.To4() != nil {
+		// Apply mask to get first address in subnet. Use To4 to ensure the IP is in the correct form for use with big.Int.
+		networkAddress = ipnet.IP.To4().Mask(ipnet.Mask)
+	} else {
+		networkAddress = ipnet.IP.To16().Mask(ipnet.Mask)
+	}
+
+	// If the subnet only has one address, or if the subnet is IPv4 and has two addresses return the network address.
+	if subnetExponent == 0 || (subnetExponent == 1 && ipnet.IP.To4() != nil) {
+		return networkAddress, nil
+	}
+
+	// If the subnet is IPv6 and has two addresses, return the second one (no broadcast address for IPv6).
+	if subnetExponent == 1 {
+		return big.NewInt(0).Add(big.NewInt(0).SetBytes(networkAddress), big.NewInt(1)).Bytes(), nil
+	}
+
+	subnetSize := big.NewInt(0).Exp(big.NewInt(2), big.NewInt(int64(subnetExponent)), nil)
+	networkAddressInt := big.NewInt(0).SetBytes(networkAddress)
+
+	// For Ipv4 we want to exclude the network address and the broadcast address, so we'll reduce the range we select
```

Comment:
IPv4 not Ipv4 (like IPv6 in the above comment).

Reply:
Can we do this calculation as part of the if block above on line 662, this is similar to my approach:

```go
	// Calculate how many host IPs are available in the subnet.
	usableHostsBig := big.NewInt(0).Exp(big.NewInt(2), big.NewInt(int64(size-ones)), nil)

	// Standardise input.
	ip4 := subnet.IP.To4()
	if ip4 != nil {
		subnet.IP = ip4.Mask(subnet.Mask)
		usableHostsBig.Sub(usableHostsBig, big.NewInt(2)) // Remove network and broascast address.

	} else {
		subnet.IP = subnet.IP.To16().Mask(subnet.Mask)
		usableHostsBig.Sub(usableHostsBig, big.NewInt(1)) // Remove network address (ipv6 has no broadcast address).
	}
```

https://github.com/canonical/lxd/pull/13379/commits/33c3c73cf26cfcf263677242609d914601a845fc#r1580787239

As it keeps the version specific logic in one place.